custom/icds/rules/custom_actions.py

```diff
@@ -0,0 +1,114 @@
+import pytz
+import uuid
+from casexml.apps.case.mock import CaseBlock
+from corehq.apps.data_interfaces.models import CaseRuleActionResult, AUTO_UPDATE_XMLNS
+from corehq.apps.hqcase.utils import submit_case_blocks, update_case
+from corehq.apps.users.util import SYSTEM_USER_ID
+from corehq.util.timezones.conversions import ServerTime
+from datetime import datetime
+from xml.etree import ElementTree
+
+
+def _create_tech_issue_delegate_for_escalation(tech_issue, owner_id):
+    case_id = uuid.uuid4().hex
+    caseblock = CaseBlock(
+        case_id,
+        case_type='tech_issue_delegate',
+        create=True,
+        update={'change_in_level': '1'},
+        user_id=SYSTEM_USER_ID,
+        owner_id=owner_id,
+        case_name=tech_issue.name,
+        index={'parent': (tech_issue.type, tech_issue.case_id, 'child')},
+    )
+    return submit_case_blocks(
+        ElementTree.tostring(caseblock.as_xml()),
+        tech_issue.domain,
+        user_id=SYSTEM_USER_ID,
+        xmlns=AUTO_UPDATE_XMLNS,
+    )
+
+
+def _update_existing_tech_issue_delegate(tech_issue_delegate):
+    if tech_issue_delegate.get_case_property('change_in_level') == '1':
+        change_in_level = '2'
+    else:
+        change_in_level = '1'
+
+    return update_case(
+        tech_issue_delegate.domain,
+        tech_issue_delegate.case_id,
+        case_properties={'change_in_level': change_in_level},
+        close=False,
+        xmlns=AUTO_UPDATE_XMLNS,
+    )
+
+
+def _update_tech_issue_for_escalation(case, escalated_ticket_level):
+    today = ServerTime(datetime.utcnow()).user_time(pytz.timezone('Asia/Kolkata')).done().date()
+
+    return update_case(
+        case.domain,
+        case.case_id,
+        case_properties={
+            'ticket_level': escalated_ticket_level,
+            'change_in_level': '1',
+            'touch_case_date': today.strftime('%Y-%m-%d'),
+        },
+        close=False,
+        xmlns=AUTO_UPDATE_XMLNS,
+    )
+
+
+def _get_escalated_tech_issue_delegate(tech_issue, escalated_location_id):
+    for subcase in tech_issue.get_subcases(index_identifier='parent'):
+        if (
+            subcase.type == 'tech_issue_delegate' and
+            subcase.owner_id == escalated_location_id and
+            not subcase.closed
+        ):
+            return subcase
+
+    return None
+
+
+def escalate_tech_issue(case, rule):
+    if case.type != 'tech_issue':
+        return CaseRuleActionResult()
```

Comment:
this isn't raising an exception - all we're doing is just returning an empty result to the calling process so that it knows no updates were made. this is a scenario that shouldn't happen but we want to make sure we're only updating tech_issue cases, and if not we just do nothing